### ISSUE TITLE
Extensions should be able to trigger MissingTypeException, AbortProcessingException

### ISSUE THREAD

**Issue by facboy** [google/auto#316]
I think these would be useful as extensions may want to be able to defer or abort processing of a particular type.

**Comment by eamonnmcmanus** [comment]
I'm inclined to agree, but I think we'd need to see a motivating example.

**Comment by facboy** [comment]
Hmm...well, for various reasons I've written an extension that generates a builder that's more tailored to my use case.  In much the same way that the built-in 'builder' code aborts at times (mostly invalid/incompatible signatures), I'd like to be able to report an error and abort processing of the given class.  I can throw a RuntimeException of course, but that ends up with a nasty-looking stack trace which probably isn't the appropriate behaviour.

At the moment I can't really think of specific use-case for MissingTypeException.

**Comment by eamonnmcmanus** [comment]
Well, you can emit an error message using context.processingEnvironment().getMessager(). That doesn't allow you to abort processing of the current class, but do you actually want to do that?

**Comment by facboy** [comment]
I certainly want to stop processing my extension.  So I could throw an error from somewhere, catch it in the top-level generateClass(), but if I return null or empty string, AutoValue complains that no source code was returned and I get an additional error from that.

**Comment by eamonnmcmanus** [comment]
I see. As a workaround you could obviously just generate a dummy class that conforms to the requirements (has required superclass, has required constructor), but that's obviously clunky. Perhaps there should just be a specific exception that you can throw from an extension. AutoValueProcessor would catch it and possibly emit an error message like "AutoValue extension FooExtension failed: blargle blargle", where "blargle blargle" is the message text of the exception.

**Comment by facboy** [comment]
Yeah that would work.  Possibly provide the ability to specify the element that failed too, like the existing ErrorReporter.  It would save the extension having to report its own error, unless it wanted to do something really funky that was outside the scope of the basic error reporting.

**Comment by grandstaish** [comment]
My specific use case for MissingTypeException was this: 

I have another annotation processor that collects and parses information about various annotated elements. My AutoValue extension is meant to use the output from the other processor. Since I don't think I can force my processor to always run before the AutoValue processor(?), that would mean that sometimes when AutoValue is processing a type, it won't be ready for me to process the output of my extension yet. I'd like to defer processing of that type until the next round in that situation.

**Comment by eamonnmcmanus** [comment]
OK. Throwing MissingTypeException doesn't feel like a very clean API, but we could add a method to AutoValueExtension that would have the same effect, say `signalIncompleteTypes()`.
